manager/allocator/allocator_new.go

```diff
@@ -0,0 +1,956 @@
+						continue
+					}
+					nodes = append(nodes, node)
+				}
+			}
+
+			// now, get all overlay network IDs that are in use, so we can
+			// allocate them to the nodes
+			networks, err := store.FindNetworks(tx, store.All)
+			if err != nil {
+				// TODO(dperny): we can't actually proceed if there is an error,
+				// because then we might deallocate all nodes' attachments
+			}
+			for _, network := range networks {
+				if network.DriverState != nil && network.DriverState.Name == "overlay" {
+					overlayNetworks[network.ID] = struct{}{}
+				}
+			}
+		})
+		readTxDone()
+
+		// make space for the successfully allocated nodes
+		allocatedNodes := make([]*api.Node, 0, len(nodes))
+
+		for _, node := range nodes {
+			ctx := log.WithField(ctx, "node.id", node.ID)
+			allocDone := metrics.StartTimer(allocatorActions.WithValues("node", "allocate"))
+			err := a.network.AllocateNode(node, overlayNetworks)
+			allocDone()
+
+			switch {
+			case err == nil:
+				// first, if no error occurred, then we can add this to the list of
+				// networks we will commit
+				allocatedNodes = append(allocatedNodes, node)
+			case errors.IsErrAlreadyAllocated(err):
+				// if the network is already allocated, there is nothing to do. log
+				// that and do not add this network to the batch
+				log.G(ctx).Debug("node already fully allocated")
+			case errors.IsErrRetryable(err):
+				// if the error is retryable, then we should log that error and
+				// re-add this network to our pending networks
+				log.G(ctx).WithError(err).Error("node could not be allocated, but will be retried")
+				a.pendingNodes[node.ID] = struct{}{}
+			default:
+				// default covers any other error case. specifically, error that
+				// are not retryable. for these, we should fail and not re-add them
+				// to the pending map. they will never succeed
+				log.G(ctx).WithError(err).Error("node cannot be allocated")
+			}
+		}
+
+		// if any nodes successfully allocated, commit them
+		if len(allocatedNodes) > 0 {
+			if err := a.store.Batch(func(batch *store.Batch) error {
+				for _, node := range allocatedNodes {
+					writeTxDone := metrics.StartTimer(storeLockHeld.WithValues("node", "allocate", "write"))
+					if err := batch.Update(func(tx store.Tx) error {
+						currentNode := store.GetNode(tx, node.ID)
+						if currentNode == nil {
+							// if there is no node, then it must have been
+							// deleted. deallocate our changes.
+
+							// TODO(dperny): IMPORTANT! there is a race
+							// condition here which is very dangerous! calls to
+							// deallocate are not idempotent! this means that
+							// we can't be sure that we won't inadvertently
+							// free some state when we deallocate a second
+							// time! a classic double-free case!
+							log.G(ctx).WithField("node.id", node.ID).Debug("node was deleted after allocation but before committing")
+							a.network.DeallocateNode(node)
```

Comment:
the event stream will deallocate the node in its state before we performed re-allocation. however, we may have locally allocated some new resources for the node, which have not been committed, but which must be released in our local state.

i'm unsure what the correct approach to working around this problem is. probably making Deallocate idempotent. but i would really appreciate suggestions.

Reply:
Ah, so the node object in the event isn't the same as the node object we change during allocation? If I understand correctly then, the problem case is:

1. The batch thread starts allocating things for an updated node.
2. The node is deleted. The event handler gets an event with the old state of the node. It waits for the lock.
3. The batch thread notices that the node is gone and deallocates it, from the correct (newly-allocated) state. It releases the lock.
4. The event handler gets the lock and deallocates again, from the old state.

If we get here then:

1. The node hadn't been deleted when we looked it up earlier in `processPendingAllocations`, which must have happened after we took `pendingMu`.
2. It is deleted now. So, a delete event must be on its way.
3. The event handler can't have deallocated yet, because it needs `pendingMu`.

So I guess we could keep a set of "already-deallocated" objects. We add the object to the set here to indicate that we already freed it. When the event thread later gets the lock, it will see that the deallocation was already done and not do it again. It will remove it from the "already-deallocated" set. A bit messy, but it should work.

Other solutions:
1. Ensure that we only have one object in memory at once representing the state of the node. Have the event thread deallocate in all cases. It will always have up-to-date information.
2. Use a single database transaction for `processPendingAllocations`. Then the user can't delete the node until we're done. Would this be a performance problem?